packages/rx_bloc_cli/mason_templates/bricks/rx_bloc_base/__brick__/lib/base/routers/router.dart

```diff
@@ -1,17 +1,232 @@
-    MaterialRoute(page: LoginPage),
-    MaterialRoute(page: NotificationsPage),
+import 'package:flutter/material.dart';
+import 'package:go_router/go_router.dart';
+import 'package:provider/provider.dart';
+
+import '../../feature_counter/di/counter_page_with_dependencies.dart';
+import '../../feature_enter_message/di/enter_message_with_dependencies.dart';
+import '../../feature_item_details/di/item_details_with_dependencies.dart';
+import '../../feature_items_list/di/items_list_with_dependencies.dart';
+import '../../feature_login/di/login_page_with_dependencies.dart';
+import '../../feature_notifications/di/notifications_page_with_dependencies.dart';
+import '../../feature_splash/di/splash_page_with_dependencies.dart';
+import '../common_blocs/coordinator_bloc.dart';
+import '../common_services/permissions_service.dart';
+import '../models/errors/error_model.dart';
+import '../models/item_model.dart';
+import 'go_router_refresh_stream.dart';
+import 'router_paths.dart';
+import 'router_permission_names.dart';
+
+part 'router.g.dart';
+
+class AppRouter {
+  AppRouter(this._context) {
+    refreshListener = GoRouterRefreshStream(
+        _context.read<CoordinatorBlocType>().states.isAuthenticated);
+  }
+
+  final BuildContext _context;
+  late GoRouterRefreshStream refreshListener;
+
+  GoRouter get router => _goRouter;
+
+  late final GoRouter _goRouter = GoRouter(
+    initialLocation: const SplashRoute().location,
+    routes: $appRoutes,
+    redirect: (context, state) async {
+      if (refreshListener.isLoggedIn && state.queryParams['from'] != null) {
+        return state.queryParams['from'];
+      }
+      if (refreshListener.isLoggedIn &&
+          state.subloc == const LoginRoute().location) {
+        return const CounterRoute().location;
+      }
+
+      final pathInfo =
+      router.routeInformationParser.matcher.findMatch(state.location);
+
+      final routeName = RouterPaths.pathToRouteName(pathInfo.fullpath);
+      final hasPermissions =
+      await context.read<PermissionsService>().hasPermission(routeName);
+
+      if (!refreshListener.isLoggedIn && !hasPermissions) {
+        return '${const LoginRoute().location}?from=${state.location}';
+      }
+
+      if (!hasPermissions) {
+        throw AccessDeniedErrorModel();
+      }
+
+      return null;
+    },
+    refreshListenable: GoRouterRefreshStream(
+      _context.read<CoordinatorBlocType>().states.isAuthenticated,
+    ),
+    errorPageBuilder: (context, state) => MaterialPage(
+      key: state.pageKey,
+      child: Scaffold(
+        appBar: AppBar(
+          title: const Text('Error page!'),
+        ),
+        body: const Center(
+          child: Text('Error message.'),
+        ),
+      ),
+    ),
+  );
+}
+
+abstract class RouteData {
+  String get permissionName;
+  String get routeLocation;
+}
+
+@TypedGoRoute<SplashRoute>(path: '/splash')
+@immutable
+class SplashRoute extends GoRouteData implements RouteData {
```

Comment:
Do you think is good idea to separate this file using the part/part of approach ?

Reply:
Done